### PR TITLE
Implement on-the-fly binning

### PR DESCRIPTION
**Pull request by fkrinner**
Implemented on the fly binning. The structure of the event file members will be changed to tuples and a check for the right ordering of eventFiles will be added, that amplitudes for different waves can't be mixed up.

### CODE REVIEW

File: partialWaveFit/pwaLikelihood.cc [hunk 1]
```diff
@@ -1089,8 +1098,23 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
 		return false;
 	}
-
-	const string waveName = meta.objectBaseName();
+	if (metas.size() == 0) {
+		printErr << "no amplitudeMetadata pointer given" << endl;
+		return false;
+	}
+	for (size_t nMeta = 0; nMeta < metas.size(); ++nMeta) {
+		if (metas[nMeta] == 0) {
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: Better: `if (not metas[nMeta]) {`

File: partialWaveFit/pwaLikelihood.cc [hunk 2]
```diff
@@ -1101,33 +1125,86 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
-	for (long int eventIndex = 0; eventIndex < meta.amplitudeTree()->GetEntriesFast(); ++eventIndex) {
-		meta.amplitudeTree()->GetEntry(eventIndex);
-		if (!ampTreeLeaf) {
-			printWarn << "null pointer to amplitude leaf for event " << eventIndex << ". "
-			          << "skipping." << endl;
-			continue;
+	bool onTheFlyBinning = false;
+	if (_eventFileProperties.size() > 0) {
+		onTheFlyBinning = true;
+	}
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: You could do `bool onTheFlyBinning = not _eventFileProperties.empty()`

File: partialWaveFit/pwaLikelihood.cc [hunk 3]
```diff
@@ -1101,33 +1125,86 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
+		size_t skipEvents  = 0;
+		size_t nmbEvtFiles = 1;
+		if (onTheFlyBinning) {
+			nmbEvtFiles = meta->eventMetadata().size();
+		}
+		for (size_t iEvt = 0; iEvt < nmbEvtFiles; ++iEvt) {
+			std::string evtHash = "";
+			if (onTheFlyBinning) {
+				evtHash = meta->eventMetadata()[iEvt].contentHash();
+				if (_eventFileHashOrder.size() < runningEventFileIndex+1){
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: What if I add more amplitudes for the second wave than for the first? That would not be caught by this, would it? It should be, should it not?

File: partialWaveFit/pwaLikelihood.cc [hunk 4]
```diff
@@ -1101,33 +1125,86 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
+					_eventFileHashOrder.push_back(evtHash);
+				} else {
+					if (_eventFileHashOrder[runningEventFileIndex] != evtHash){
+						printErr << "order of eventFiles differs between amplitude. Order of the events mixed up. Aborting..." << std::endl;
+						return false;
+					}
+				}
+				++runningEventFileIndex;
+			}
+			size_t startIndex = 0;
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: Do you need the `startIndex` anywhere?

File: partialWaveFit/pwaLikelihood.cc [hunk 5]
```diff
@@ -1101,33 +1125,86 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
+			size_t stopIndex  = meta->amplitudeTree()->GetEntriesFast();
+			if (onTheFlyBinning) {
+				stopIndex = _eventFileProperties[evtHash].second.size();
+			}
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: If you use the ternary `?:` operator, you could make the indices `const`, i.e.:

`const size_t stopIndex  = onTheFlyBinning ? _eventFileProperties[evtHash].second.size() : meta->amplitudeTree()->GetEntriesFast();`

File: partialWaveFit/pwaLikelihood.cc [hunk 6]
```diff
@@ -1228,6 +1304,48 @@ pwaLikelihood<complexT>::finishInit()
...
+		const rpwa::eventMetadata* evtMeta = evtMetas[metaIndex];
+		const std::string evtHash = evtMeta->contentHash();
+		if (evtMeta == 0) {
+			printErr << "eventMetadata not set. Aborting..." << endl;
+			return false;
+		}
+		std::map<std::string, double> binningVariables;
+		typedef std::map<std::string, std::pair<double, double> >::const_iterator it_type;
+		for(it_type iterator = binningMap->begin(); iterator != binningMap->end(); iterator++) {
+			std::string additionalVar = iterator->first;
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This should be `const`

File: partialWaveFit/pwaLikelihood.cc [hunk 7]
```diff
@@ -1228,6 +1304,48 @@ pwaLikelihood<complexT>::finishInit()
...
+		std::vector<size_t> eventIndices;
+		for (int eventIndex = 0; eventIndex < evtMeta->eventTree()->GetEntriesFast(); ++eventIndex) {
+			evtMeta->eventTree()->GetEntry(eventIndex);
+			bool useEvent = true;
+			typedef std::map<std::string, std::pair<double, double> >::const_iterator it_type;
+			for(it_type iterator = binningMap->begin(); iterator != binningMap->end(); iterator++) {
+				std::string additionalVar = iterator->first;
+				double lowerBound = iterator->second.first;
+				double upperBound = iterator->second.second;
+				double binVarValue = binningVariables[additionalVar];
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: There should be `const` references.

File: partialWaveFit/pwaLikelihood.cc [hunk 8]
```diff
@@ -1228,6 +1304,48 @@ pwaLikelihood<complexT>::finishInit()
...
+		}
+		std::map<std::string, double> binningVariables;
+		typedef std::map<std::string, std::pair<double, double> >::const_iterator it_type;
+		for(it_type iterator = binningMap->begin(); iterator != binningMap->end(); iterator++) {
+			std::string additionalVar = iterator->first;
+			binningVariables.insert(std::pair<string, double>(additionalVar, 0.));
+			evtMeta->eventTree()->SetBranchAddress(additionalVar.c_str(), &binningVariables[additionalVar]);
+		}
+		std::vector<size_t> eventIndices;
+		for (int eventIndex = 0; eventIndex < evtMeta->eventTree()->GetEntriesFast(); ++eventIndex) {
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: `eventIndex` should be a `long`.

File: partialWaveFit/pwaLikelihood.cc [hunk 9]
```diff
@@ -1153,7 +1230,6 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
 	return true;
 }
 
-
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: Please re-add this line.

File: partialWaveFit/pwaLikelihood.h [hunk 10]
```diff
@@ -180,9 +182,12 @@ namespace rpwa {
 		bool addAccIntegral(rpwa::ampIntegralMatrix& accMatrix, unsigned int accEventsOverride = 0);
 
 		bool addAmplitude(const rpwa::amplitudeMetadata& meta);
+		bool addAmplitude(const std::vector<const rpwa::amplitudeMetadata*> meta);
 
 		bool finishInit();
 
+		bool setOnTheFlyBinning(const std::map<std::string, std::pair<double, double> >* binningMap, const std::vector<const eventMetadata*> evtMeta);
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.h]: Could you write the arguments on separate lines, as with the `getIntegralMatrices` below?

File: pyInterface/package/_likelihood.py [hunk 11]
```diff
@@ -1,65 +1,93 @@
...
+                   accIntegralFileName, 
+                   binningMap, 
+                   waveDescThres, 
+                   cauchy            = False, 
+                   cauchyWidth       = 0.5,  
+                   accEventsOverride = 0, 
+                   rank              = 1, 
+                   verbose           = False, 
+                   eventFileNameList   = []
+                   ):
```
Comment (legordian) [inline pyInterface/package/_likelihood.py]: Is there any pressing reason to re-order the whole argument list?

File: pyInterface/package/_fit.py [hunk 12]
```diff
@@ -90,18 +92,17 @@ def pwaNloptFit(ampFileList,
...
-
-	likelihood = pyRootPwa.initLikelihood(waveDescThres = waveDescThres,
-	                                      massBinCenter = massBinCenter,
-	                                      ampFileList = ampFileList,
+	likelihood = pyRootPwa.initLikelihood(ampFileList = ampFileList,
 	                                      normIntegralFileName = normIntegralFileName,
 	                                      accIntegralFileName = accIntegralFileName,
-	                                      accEventsOverride = accEventsOverride,
+	                                      binningMap = binningMap,
+                                              waveDescThres = waveDescThres,
```
Comment (legordian) [inline pyInterface/package/_fit.py]: Indentation is wrong.

File: pyInterface/package/_likelihood.py [hunk 13]
```diff
@@ -1,65 +1,93 @@
...
+		pyRootPwa.utils.printErr("could not initialize likelihood. Aborting...")
 		return None
-
-	normIntFile = ROOT.TFile.Open(normIntegralFileName, "READ")
-	if len(normIntFile.GetListOfKeys()) != 1:
-		pyRootPwa.utils.printWarn("'" + normIntegralFileName + "' does not contain exactly one TKey.")
+ 
+ 	normIntFile = ROOT.TFile.Open(normIntegralFileName, "READ")
+ 	if len(normIntFile.GetListOfKeys()) != 1:
+ 		pyRootPwa.utils.printWarn("'" + normIntegralFileName + "' does not contain exactly one TKey.")
```
Comment (legordian) [inline pyInterface/package/_likelihood.py]: Could you fix the indentation? There are extra spaces at the beginning of the lines (does the python interpreter even allow this?)...

File: pyInterface/package/_likelihood.py [hunk 14]
```diff
@@ -1,65 +1,93 @@
...
-		if (not likelihood.addAmplitude(meta)):
-			pyRootPwa.utils.printErr("could not add amplitude '" + waveName + "'. Aborting...")
+		eventMetas.append(evtMeta)
+		openFiles.append(evtFile)
+	likelihood.setOnTheFlyBinning(binningMap, eventMetas)
+ 	for wave in waveDescThres:
+		openFilesWave = []
+ 		waveName = wave[0]
+		ampFileNameList = ampFileList[waveName]
+		if not len(ampFileNameList) == len(eventMetas) and len(eventMetas) > 0:
```
Comment (legordian) [inline pyInterface/package/_likelihood.py]: Other way around makes more sense, i.e.:

`if len(eventMetas) > 0 and not len(ampFileNameList) == len(eventMetas):`

File: pyInterface/partialWaveFit/pwaLikelihood_py.cc [hunk 15]
```diff
@@ -195,6 +213,38 @@ namespace {
 		return bp::list(self.CorrectParamSigns(par.data()));
 	}
 
+	bool 
+	pwaLikelihood_setOnTheFlyBinning(rpwa::pwaLikelihood<std::complex<double> >& self,
+                                         bp::dict                                    pyBinningMap,
```
Comment (legordian) [inline pyInterface/partialWaveFit/pwaLikelihood_py.cc]: Indentation is wrong.

File: pyInterface/partialWaveFit/pwaLikelihood_py.cc [hunk 16]
```diff
@@ -195,6 +213,38 @@ namespace {
...
+		if (not rpwa::py::convertBPObjectToVector<const rpwa::eventMetadata*>(pyEvtMetas, evtMetas)){
+			PyErr_SetString(PyExc_TypeError, "could not extract event metadatas");
+			bp::throw_error_already_set();
+		}
+		return self.setOnTheFlyBinning(binningMap, evtMetas);
+	}
+
+	bool
+	pwaLikelihood_addAmplitude4(rpwa::pwaLikelihood<std::complex<double> >& self,
+	                             bp::list                                   pyMetas)
```
Comment (legordian) [inline pyInterface/partialWaveFit/pwaLikelihood_py.cc]: Indentation is wrong.

File: pyInterface/partialWaveFit/pwaLikelihood_py.cc [hunk 17]
```diff
@@ -195,6 +213,38 @@ namespace {
...
+		std::vector<const rpwa::eventMetadata*> evtMetas; 
+		if (not rpwa::py::convertBPObjectToVector<const rpwa::eventMetadata*>(pyEvtMetas, evtMetas)){
+			PyErr_SetString(PyExc_TypeError, "could not extract event metadatas");
+			bp::throw_error_already_set();
+		}
+		return self.setOnTheFlyBinning(binningMap, evtMetas);
+	}
+
+	bool
+	pwaLikelihood_addAmplitude4(rpwa::pwaLikelihood<std::complex<double> >& self,
```
Comment (legordian) [inline pyInterface/partialWaveFit/pwaLikelihood_py.cc]: Why is this here? It is not used anywhere?

File: partialWaveFit/pwaLikelihood.h [hunk 18]
```diff
@@ -49,6 +50,7 @@
 
 #include "sumAccumulators.hpp"
 #include "ampIntegralMatrix.h"
+#include "eventMetadata.h"
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.h]: Forward-declaration of eventMetadata should be enough, should it not?

File: partialWaveFit/pwaLikelihood.h [hunk 19]
```diff
@@ -180,9 +182,13 @@ namespace rpwa {
 		bool addAccIntegral(rpwa::ampIntegralMatrix& accMatrix, unsigned int accEventsOverride = 0);
 
 		bool addAmplitude(const rpwa::amplitudeMetadata& meta);
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.h]: Could you please remove this version of the interface and propagate that throughout the project.

File: partialWaveFit/pwaLikelihood.h [hunk 20]
```diff
@@ -180,9 +182,13 @@ namespace rpwa {
...
+		bool addAmplitude(const std::vector<const rpwa::amplitudeMetadata*> meta);
 
 		bool finishInit();
 
+		bool setOnTheFlyBinning(const std::map<std::string, std::pair<double, double> >* binningMap,
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.h]: Reference instead of pointer would be better as a type for the binningMap.

File: partialWaveFit/pwaLikelihood.h [hunk 21]
```diff
@@ -273,6 +279,8 @@ namespace rpwa {
 
 		mutable functionCallInfo _funcCallInfo[NMB_FUNCTIONCALLENUM];  // collects function call statistics
 
+		std::vector<std::string> _eventFileHashOrder;
+		std::map<std::string,std::pair<size_t, std::vector<size_t> > > _eventFileProperties;
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.h]: There should be a space after the first `std::string`.

File: partialWaveFit/pwaLikelihood.cc [hunk 22]
```diff
@@ -1089,8 +1098,23 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
+		printErr << "no amplitudeMetadata pointer given" << endl;
+		return false;
+	}
+	for (size_t nMeta = 0; nMeta < metas.size(); ++nMeta) {
+		if (not metas[nMeta]) {
+			printErr << "amplitudeMetadata metas[" << nMeta << "] not valid. Aborting..." << endl;
+			return false;
+		}
+	}
+	const string waveName = metas[0]->objectBaseName();
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This could be a reference.

File: partialWaveFit/pwaLikelihood.cc [hunk 23]
```diff
@@ -1089,8 +1098,23 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
+	for (size_t nMeta = 0;nMeta < metas.size(); ++nMeta) {
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: nMeta could start at 1, also, there is a space missing after the first semicolon.

File: partialWaveFit/pwaLikelihood.cc [hunk 24]
```diff
@@ -1101,33 +1125,77 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
-	amplitudeTreeLeaf* ampTreeLeaf = 0;
-	meta.amplitudeTree()->SetBranchAddress(amplitudeMetadata::amplitudeLeafName.c_str(), &ampTreeLeaf);
-	vector<complexT> amps;
-	for (long int eventIndex = 0; eventIndex < meta.amplitudeTree()->GetEntriesFast(); ++eventIndex) {
-		meta.amplitudeTree()->GetEntry(eventIndex);
-		if (!ampTreeLeaf) {
-			printWarn << "null pointer to amplitude leaf for event " << eventIndex << ". "
-			          << "skipping." << endl;
-			continue;
+	bool onTheFlyBinning = not _eventFileProperties.empty();
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This should be `const`.

File: partialWaveFit/pwaLikelihood.cc [hunk 25]
```diff
@@ -1101,33 +1125,77 @@ pwaLikelihood<complexT>::addAmplitude(const amplitudeMetadata& meta)
...
+	size_t totalEvents = 0;
+	for (size_t nMeta = 0; nMeta < metas.size(); ++nMeta) {
+		if (onTheFlyBinning) {
+			vector<eventMetadata> evtMetas = metas[nMeta]->eventMetadata();
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This should be a `const` reference.

File: partialWaveFit/pwaLikelihood.h [hunk 26]
```diff
@@ -179,10 +181,13 @@ namespace rpwa {
 
 		bool addAccIntegral(rpwa::ampIntegralMatrix& accMatrix, unsigned int accEventsOverride = 0);
 
-		bool addAmplitude(const rpwa::amplitudeMetadata& meta);
+		bool addAmplitude(const std::vector<const rpwa::amplitudeMetadata*> meta);
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.h]: A `const` reference to the vector would probably be better.

File: partialWaveFit/pwaLikelihood.cc [hunk 27]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
 
 template<typename complexT>
 bool
+pwaLikelihood<complexT>::setOnTheFlyBinning(const map<string, pair<double, double> >& binningMap,
+                                            const vector<const eventMetadata*>        evtMetas)
+{
+	for (size_t metaIndex = 0; metaIndex < evtMetas.size(); ++metaIndex) {
+		const eventMetadata* evtMeta = evtMetas[metaIndex];
+		const string evtHash = evtMeta->contentHash();
+		if (evtMeta == 0) {
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This check should move up, and also look like `(not evtMeta)`.

File: partialWaveFit/pwaLikelihood.cc [hunk 28]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
 
 template<typename complexT>
 bool
+pwaLikelihood<complexT>::setOnTheFlyBinning(const map<string, pair<double, double> >& binningMap,
+                                            const vector<const eventMetadata*>        evtMetas)
+{
+	for (size_t metaIndex = 0; metaIndex < evtMetas.size(); ++metaIndex) {
+		const eventMetadata* evtMeta = evtMetas[metaIndex];
+		const string evtHash = evtMeta->contentHash();
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This should be a reference.

File: partialWaveFit/pwaLikelihood.cc [hunk 29]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+			string additionalVar = iterator->first;
+			binningVariables.insert(pair<string, double>(additionalVar, 0.));
+			evtMeta->eventTree()->SetBranchAddress(additionalVar.c_str(), &binningVariables[additionalVar]);
+		}
+		vector<size_t> eventIndices;
+		for (long eventIndex = 0; eventIndex < evtMeta->eventTree()->GetEntriesFast(); ++eventIndex) {
+			evtMeta->eventTree()->GetEntry(eventIndex);
+			bool useEvent = true;
+			typedef map<string, pair<double, double> >::const_iterator it_type;
+			for(it_type iterator = binningMap.begin(); iterator != binningMap.end(); iterator++) {
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: `++iterator`!

File: partialWaveFit/pwaLikelihood.cc [hunk 30]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+	for (size_t metaIndex = 0; metaIndex < evtMetas.size(); ++metaIndex) {
+		const eventMetadata* evtMeta = evtMetas[metaIndex];
+		const string evtHash = evtMeta->contentHash();
+		if (evtMeta == 0) {
+			printErr << "eventMetadata not set. Aborting..." << endl;
+			return false;
+		}
+		map<string, double> binningVariables;
+		typedef map<string, pair<double, double> >::const_iterator it_type;
+		for(it_type iterator = binningMap.begin(); iterator != binningMap.end(); iterator++) {
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: `++iterator`!

File: partialWaveFit/pwaLikelihood.cc [hunk 31]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+			string additionalVar = iterator->first;
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This should be a `const` reference.

File: partialWaveFit/pwaLikelihood.cc [hunk 32]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+			binningVariables.insert(pair<string, double>(additionalVar, 0.));
+			evtMeta->eventTree()->SetBranchAddress(additionalVar.c_str(), &binningVariables[additionalVar]);
+		}
+		vector<size_t> eventIndices;
+		for (long eventIndex = 0; eventIndex < evtMeta->eventTree()->GetEntriesFast(); ++eventIndex) {
+			evtMeta->eventTree()->GetEntry(eventIndex);
+			bool useEvent = true;
+			typedef map<string, pair<double, double> >::const_iterator it_type;
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: This line is not needed.

File: partialWaveFit/pwaLikelihood.cc [hunk 33]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+			for(it_type iterator = binningMap.begin(); iterator != binningMap.end(); iterator++) {
+				const string& additionalVar = iterator->first;
+				const double& lowerBound = iterator->second.first;
+				const double& upperBound = iterator->second.second;
+				const double& binVarValue = binningVariables[additionalVar];
+				if (binVarValue < lowerBound or binVarValue > upperBound) {
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: `binVarValue >= upperBound`

File: partialWaveFit/pwaLikelihood.cc [hunk 34]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+					useEvent = false;
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: Insert `break`.

File: partialWaveFit/pwaLikelihood.cc [hunk 35]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+				}
+			}
+			if (useEvent) {
+					eventIndices.push_back(eventIndex);
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: There is a tab too much in this line.

File: partialWaveFit/pwaLikelihood.cc [hunk 36]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+			}
+		}
+		printInfo << "found " << eventIndices.size() << " of " << evtMeta->eventTree()->GetEntriesFast() << " in the event file to be in given bin" << endl;
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: Add period at the end of output.

File: partialWaveFit/pwaLikelihood.cc [hunk 37]
```diff
@@ -1231,6 +1290,50 @@ pwaLikelihood<complexT>::finishInit()
...
+		_eventFileProperties[evtHash] = pair<size_t,vector<size_t> >(evtMeta->eventTree()->GetEntriesFast(), eventIndices);
```
Comment (legordian) [inline partialWaveFit/pwaLikelihood.cc]: Add space after `size_t`.

File: pyInterface/partialWaveFit/pwaLikelihood_py.cc [hunk 38]
```diff
@@ -55,6 +55,18 @@ namespace {
 
 
 	bool
+	pwaLikelihood_addAmplitude(rpwa::pwaLikelihood<std::complex<double> >& self,
+	                             bp::list                                   pyMetas)
```
Comment (legordian) [inline pyInterface/partialWaveFit/pwaLikelihood_py.cc]: Indendation problems.

File: pyInterface/partialWaveFit/pwaLikelihood_py.cc [hunk 39]
```diff
@@ -55,6 +55,18 @@ namespace {
...
+	{
+		std::vector<const rpwa::amplitudeMetadata*> metas;
+		if (not rpwa::py::convertBPObjectToVector<const rpwa::amplitudeMetadata*>(pyMetas, metas)){
+			PyErr_SetString(PyExc_TypeError, "could not extract amplitudes metadatas");
```
Comment (legordian) [inline pyInterface/partialWaveFit/pwaLikelihood_py.cc]: `"could not extract vector of amplitude metadata"`

File: pyInterface/partialWaveFit/pwaLikelihood_py.cc [hunk 40]
```diff
@@ -55,6 +55,18 @@ namespace {
...
+			bp::throw_error_already_set();
+		};
```
Comment (legordian) [inline pyInterface/partialWaveFit/pwaLikelihood_py.cc]: Remove `;`.

File: pyInterface/partialWaveFit/pwaLikelihood_py.cc [hunk 41]
```diff
@@ -195,6 +207,26 @@ namespace {
 		return bp::list(self.CorrectParamSigns(par.data()));
 	}
 
+	bool
+	pwaLikelihood_setOnTheFlyBinning(rpwa::pwaLikelihood<std::complex<double> >& self,
+	                                 bp::dict                                    pyBinningMap,
+	                                 bp::list                                    pyEvtMetas)
+	{
+		std::map<std::string, std::pair<double, double> > binningMap;
+		bp::list keys = pyBinningMap.keys();
```
Comment (legordian) [inline pyInterface/partialWaveFit/pwaLikelihood_py.cc]: This should be `const`.

File: legacy/pwaNloptFit.cc [hunk 42]
```diff
@@ -361,7 +361,7 @@ main(int    argc,
 			exit(1);
 		}
 
-		if (not L.addAmplitude(*ampMeta)) {
+		if (not L.addAmplitude(std::vector<const amplitudeMetadata*>(1,ampMeta))) {
```
Comment (legordian) [inline legacy/pwaNloptFit.cc]: Space missing after `,`.

File: legacy/pwafit.cc [hunk 43]
```diff
@@ -361,7 +361,7 @@ main(int    argc,
 			exit(1);
 		}
 
-		if (not L.addAmplitude(*ampMeta)) {
+		if (not L.addAmplitude(std::vector<const amplitudeMetadata*>(1,ampMeta))) {
```
Comment (legordian) [inline legacy/pwafit.cc]: Space missing after `,`.

File: pyInterface/package/_likelihood.py [hunk 44]
```diff
@@ -43,23 +47,48 @@ def initLikelihood(waveDescThres,
...
-			pyRootPwa.utils.printErr("could not open amplitude file '" + ampFileName + "'.")
+		ampFileNameList = ampFileList[waveName]
+		if len(eventMetas) > 0 and not len(ampFileNameList) == len(eventMetas):
+			pyRootPwa.utils.printErr("number of amplitude and event files do not match")
 			return None
-		meta = pyRootPwa.core.amplitudeMetadata.readAmplitudeFile(ampFile, waveName)
-		if not meta:
-			pyRootPwa.utils.printErr("could not get metadata for waveName '" + waveName + "'.")
-			return None
-		if (not likelihood.addAmplitude(meta)):
```
Comment (legordian) [inline pyInterface/package/_likelihood.py]: Remove all changes in .py files, instead change this line to
`if (not likelihood.addAmplitude( [meta] )):`

**Comment by suhlatwork** [comment]
Merged after another (small) round of coding style fixes and a rebase as 74d8c9f3f3ad5f5589bef2e8101aa09d40bbb643.
